### PR TITLE
Migrate panel stimulus w panel

### LINKED ISSUE #10168: 🎛️ Migrate `panels` to a Stimulus controller `w-panel`
> ℹ️ **Part of the [Stimulus 🎛️ RFC 78](https://github.com/wagtail/rfcs/pull/78)**

### Is your proposal related to a problem?

At the moment we have an ad-hoc TypeScript implementation for expanding panels used extensively in the Wagtail admin.

It would be great to have this implementation migrated to Stimulus so that it can easily be attached to dynamically injected HTML content and be a more flexible and testable solution.

### Describe the solution you'd like

See the current implementation.

[`wagtail/admin/templates/wagtailadmin/shared/panel.html`](https://github.com/wagtail/wagtail/blob/main/wagtail/admin/templates/wagtailadmin/shared/panel.html)

[`client/src/includes/panels.ts`](https://github.com/wagtail/wagtail/blob/main/client/src/includes/panels.ts)

Proposal is to replace the `data-panel` attributes with the suitable Stimulus data attributes.

```html
<section class="w-panel ..." ...attrs data-controller="w-panel" data-action="w-panel:toggle->w-panel#toggle">
    {# If a panel has no heading nor header controls, we don’t want any of the associated UI. #}
    {% if heading or header_controls %}
        <div class="w-panel__header">
            <a class="w-panel__anchor w-panel__anchor--prefix" href="#{{ panel_id }}" aria-labelledby="{{ heading_id }}" data-w-panel-target="anchor">
                {% icon name="link" classname="w-panel__icon" %}
            </a>
            <button class="w-panel__toggle" type="button" aria-label="{% trans 'Toggle section' %}" aria-describedby="{{ heading_id }}" aria-controls="{{ content_id }}" aria-expanded="true" data-action="w-panel->toggle">
                {% firstof icon "arrow-down-big" as icon_name %}
                {% icon name=icon_name classname="w-panel__icon" %}
            </button>
            {% if heading %}
                <h2 class="w-panel__heading {% if heading_size == "label" %}w-panel__heading--label{% endif %}" {% if heading_level %}aria-level="{{ heading_level }}"{% endif %} id="{{ heading_id }}" data-w-panel-target="heading">
                    {% if id_for_label %}
                        <label for="{{ id_for_label }}" id="{{ id_for_label }}-label"><span data-w-panel-target="heading-text">{{ heading }}</span>{% if is_required %}<span class="w-required-mark" data-panel-required>*</span>{% endif %}</label>
                    {% else %}
                        <span data-w-panel-target="heading-text">{{ heading }}</span>{% if is_required %}<span class="w-required-mark" data-panel-required>*</span>{% endif %}
                    {% endif %}
                </h2>
            {% endif %}
            <a class="w-panel__anchor w-panel__anchor--suffix" href="#{{ panel_id }}" aria-labelledby="{{ heading_id }}">
                {% icon name="link" classname="w-panel__icon" %}
            </a>
            <div class="w-panel__divider"></div>
            {% if header_controls %}
                <div class="w-panel__controls" data-w-panel-target="controls">
                    <div class="w-panel__controls-cue">
                        {% icon name="dots-horizontal" classname="w-panel__icon" %}
                    </div>
                    {{ header_controls }}
                </div>
            {% endif %}
        </div>
    {% endif %}

    <div id="{{ content_id }}" class="w-panel__content" data-w-panel-target="content" data-action="beforematch->w-panel#toggle">
        {{ children }}
    </div>
</section>
```

The Stimulus controller will need to be implemented with the following features, may need some revision once development starts.

Note: Try to use the term `open` throughout, instead of a mix of `expanded/collapsed or open/closed` as it makes the logic hard to follow. Also, `open` is what the HTML [`details`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) element uses, and is easier to spell.

- targets
  - `anchor`
  - `heading`
  - `headingText`
  - `container`
- values
  - `find` (defaults to `'[aria-invalid="true"], .error, .w-field--error'`) - for finding errors / content that should force it to be open
  - `open` - state for collapsed or expanded, default is false
- methods
  - `connect` - Should smooth scroll IF the anchor matches, should check the 'find' elements to determine if it should be forced to expand and also set the hidden until-found attribute if browser support exists. Once done it should dispatch a connected event `this.dispatchEvent('connected', { detail: { expanded, collapsed? }})` once done (replaces `'wagtail:panel-init'`). Event names based on something done should be past tense.
  - `toggle` - should accept three ways to 'force' the behaviour in one direction; called with a Boolean and use that, `event.params.force` or `event.detail.force`. can use a better name, but avoid multi-word if possible. Should dispatch `w-panel:toggled` note, toggled, not toggle, when done. Important: events based on something done should be past tense.
  - `scroll` - just a split out method to scroll the 'active' panel into view, as done in `initAnchoredPanels`.

**Considerations**

- We will no longer expose a `toggleCollapsiblePanel` so any usage of that function must instead dispatch an event on the DOM element (this is why )
- Must consider all the other code usage of any of the `data-panel` attributes, including setting up in client/src/components/StreamField/blocks/BaseSequenceBlock.js with all the right attributes.
- Consider the minimap usage also.
- Ensure that any event listening is updated for any existing event names that have been renamed, likely needs a release notes entry also.
- `w-panel` may not be the best name for this, it's really about collapsing and expanding details, but the name is already in use so it might be simpler than `w-collapse` or `w-details`.
- Long term we should consider how we may use `details` element in the future.
- Review how the global expand/collapse button works and ensure that behaviour still is working as expected.

### Describe alternatives you've considered

- Leave as is, the panel code is pretty robust and already using lots of data attributes.

### Additional context

- This is part of a larger goal to migrate to Stimulus
- Not the same thing, but there is also some related work on Panel (python/Django widgets) using `attrs` #10133

### PR DESCRIPTION
**Pull request by mansi0829**
Fixes #10168

_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [x] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

Development environment: Ubuntu 22.04 64-bit, python 3.10.6

### CODE REVIEW

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
@mansi0829 thank you for giving this a shot! Could you take a moment to update the pull request’s description, so we better understand your approach and what you intend to do at this stage of the work?

**Comment by mansi0829** [comment]
> @mansi0829 thank you for giving this a shot! Could you take a moment to update the pull request’s description, so we better understand your approach and what you intend to do at this stage of the work?

i was planning to create an independent panel controller in which we can write panel toggle logic for w panel classes

**Comment by lb-** [comment]
Please note though that the code so far is not close to what's expected for a Stimulus controller approach. The data attributes added are incorrect, however, as this is a draft I assume you are still working through this and learning Stimulus. 

Let us know when you feel this is ready for some feedback.

**Comment by mansi0829** [comment]
> Please note though that the code so far is not close to what's expected for a Stimulus controller approach. The data attributes added are incorrect, however, as this is a draft I assume you are still working through this and learning Stimulus.
> 
> Let us know when you feel this is ready for some feedback.

I have changed the description and I am working on creating a controller till tomorrow I will try to give the update

**Comment by mansi0829** [comment]
@lb-  can we use this component [https://www.stimulus-components.com/docs/stimulus-reveal-controller/](url) 
here classes are making our work more easier

**Comment by lb-** [comment]
@mansi0829 great question. Answered on the original issue. 

https://github.com/wagtail/wagtail/issues/10168#issuecomment-1483563649

**Comment by mansi0829** [comment]
> @mansi0829 great question. Answered on the original issue.
> 
> [#10168 (comment)](https://github.com/wagtail/wagtail/issues/10168#issuecomment-1483563649)

<b>After changes</b>
Screenshot from 2023-03-26 20-49-32
Screenshot from 2023-03-26 20-49-08

**Comment by mansi0829** [comment]
when i make changes in ```BaseSequenceBlock``` few things are also changed with should i address this in same issue or create different pr

File: client/src/components/StreamField/blocks/BaseSequenceBlock.js [hunk 1]
```diff
@@ -143,7 +143,7 @@ export class BaseSequenceChild extends EventEmitter {
           ? `data-contentpath="${h(this.id)}"`
           : 'data-contentpath-disabled'
       }>
-        <input type="hidden"  name="${this.prefix}-deleted" value="">
+        <input type="hidden" name="${this.prefix}-deleted" value="">
```
Comment (lb-) [inline client/src/components/StreamField/blocks/BaseSequenceBlock.js:146]: Try to avoid unrelated changes, this has actually been fixed up on main I think anyway.

File: client/src/controllers/RevealController.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ *           </div>
+ *   </div>
+ *   <div
+ * w-panel-target="item"
+ * data-action="beforematch->w-panel#toggle"
+ * >
+ *   </div>
+ * </section>
+ * */
+export class RevealController extends Controller<HTMLButtonElement> {
```
Comment (lb-) [inline client/src/controllers/RevealController.ts:28]:
```suggestion
export class RevealController extends Controller<HTMLElement> {
```

The controlled element is actually the root div, so just HTMLElement here is sufficient.

File: wagtail/admin/templates/wagtailadmin/shared/panel.html [hunk 3]
```diff
@@ -19,23 +19,23 @@
 {% fragment as panel_id %}{{ prefix }}-section{% endfragment %}
 {% fragment as heading_id %}{{ prefix }}-heading{% endfragment %}
 {% fragment as content_id %}{{ prefix }}-content{% endfragment %}
-<section class="w-panel {{ classname }}" id="{{ panel_id }}" {% if heading %}aria-labelledby="{{ heading_id }}"{% endif %} data-panel>
+<section class="w-panel {{ classname }}" id="{{ panel_id }}" {% if heading %}aria-labelledby="{{ heading_id }}"{% endif %} w-panel data-controller="w-panel" >
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/panel.html:22]:
```suggestion
<section class="w-panel {{ classname }}" id="{{ panel_id }}" {% if heading %}aria-labelledby="{{ heading_id }}"{% endif %} data-controller="w-panel" >
```

We should not need the original data attributes at all once this PR is done.

File: wagtail/admin/templates/wagtailadmin/shared/panel.html [hunk 4]
```diff
@@ -19,23 +19,23 @@
...
     {# If a panel has no heading nor header controls, we don’t want any of the associated UI. #}
     {% if heading or header_controls %}
-        <div class="w-panel__header">
-            <a class="w-panel__anchor w-panel__anchor--prefix" href="#{{ panel_id }}" data-panel-anchor aria-labelledby="{{ heading_id }}">
+        <div class="w-panel__header" >
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/panel.html:25]:
```suggestion
        <div class="w-panel__header">
```
Try to avoid unrelated changes.

File: wagtail/admin/templates/wagtailadmin/shared/panel.html [hunk 5]
```diff
@@ -19,23 +19,23 @@
...
+            <a class="w-panel__anchor w-panel__anchor--prefix" href="#{{ panel_id }}" w-panel-anchor aria-labelledby="{{ heading_id }}">
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/panel.html:26]:
```suggestion
            <a class="w-panel__anchor w-panel__anchor--prefix" href="#{{ panel_id }}" data-w-panel-target="anchor" aria-labelledby="{{ heading_id }}">
```

Please note that using an attribute that is like `w-foo` is not a valid attribute. It will not break anything but should be avoided and is not necessary for this work either. 

Instead make this a target

File: wagtail/admin/templates/wagtailadmin/shared/panel.html [hunk 6]
```diff
@@ -19,23 +19,23 @@
...
                 {% icon name="link" classname="w-panel__icon" %}
             </a>
-            <button class="w-panel__toggle" type="button" aria-label="{% trans 'Toggle section' %}" aria-describedby="{{ heading_id }}" data-panel-toggle aria-controls="{{ content_id }}" aria-expanded="true">
+            <button class="w-panel__toggle" id="toggle-button" type="button" aria-label="{% trans 'Toggle section' %}" aria-describedby="{{ heading_id }}" data-w-panel-toggle data-action="click->w-panel#toggle" aria-controls="{{ content_id }}" aria-expanded="true">
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/panel.html:29]:
```suggestion
            <button class="w-panel__toggle" type="button" aria-label="{% trans 'Toggle section' %}" aria-describedby="{{ heading_id }}" data-action="w-panel#toggle" aria-controls="{{ content_id }}" aria-expanded="true">
```

Avoid setting an id on elements in this shared code, reminder that html must not have duplicate IDs. 

Similar to above, if we don't need the original panel data attributes, we should remove them. 

Finally, you can use the shorter method of declaring actions on some elements like the click event on a button.

File: client/scss/components/forms/_nested-panel.scss [hunk 7]
```diff
@@ -75,3 +75,7 @@ $panel-x-offset: calc($header-button-size-sm / 2 + $header-gap);
     display: none;
   }
 }
+
+.hidden {
```
Comment (lb-) [inline client/scss/components/forms/_nested-panel.scss:79]: I'd like us to avoid using the added class, the current implementation uses the hidden attribute. Let's try to align with exactly how the current implementation is done (JavaScript wise). 

See TS code here. https://github.com/wagtail/wagtail/blob/main/client/src/includes/panels.ts

File: wagtail/admin/templates/wagtailadmin/shared/panel.html [hunk 8]
```diff
@@ -54,7 +54,7 @@ <h2 class="w-panel__heading {% if heading_size == "label" %}w-panel__heading--la
         </div>
     {% endif %}
 
-    <div id="{{ content_id }}" class="w-panel__content">
+    <div id="toggle-target {{ content_id }}" class="w-panel__content " data-w-panel-target="item">
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/panel.html:57]:
```suggestion
    <div id="{{ content_id }}" class="w-panel__content " data-w-panel-target="item">
```

As per the original issue, let's use `container` or maybe `content` for the content that shows or hides.

**Review by lb- — CHANGES_REQUESTED** [review]
Getting closer, thanks for keeping at this. 

Please note that all attributes must be correct data attributes and some existing attributes may no longer be needed. There should not be any non-standard attributes like `w-panel` added. 

Additionally, please avoid changing element IDs. This is not needed for the implementation. 

As for the base sequence block - yes - that will need changes as per the original issue description.

File: wagtail/admin/templates/wagtailadmin/shared/panel.html (hunk 8 above)
Comment (mansi0829) [reply wagtail/admin/templates/wagtailadmin/shared/panel.html:57]: in this should i change item -> content ?

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/panel.html:57]: Yeah. The original issue has this target name specifically called out to be `container` but `content` also makes sense. 

Item is a tad too generic I think

**Comment by lb-** [comment]
@mansi0829  just checking in if you are still keen on doing this? all good if not, feel free to close - maybe others can build on your start. Or feel free to ask for feedback if you have more questions/changes applied.

File: client/scss/components/forms/_nested-panel.scss (hunk 7 above)
Comment (mansi0829) [reply client/scss/components/forms/_nested-panel.scss:79]: are you suggesting to use attribute like aria-controls bacause if i remove hidden class from scss then it isnt working

**Comment by mansi0829** [comment]
> @mansi0829 just checking in if you are still keen on doing this? all good if not, feel free to close - maybe others can build on your start. Or feel free to ask for feedback if you have more questions/changes applied.

yes i am still willing to work on this issue

**Comment by lb-** [comment]
Hey @mansi0829 - do you have any update on this? Are you still planning on pursing this PR?
Otherwise we might close this and add notes to the issue for someone else (or yourself) to pick up in the future.

**Comment by lb-** [comment]
Closing as https://github.com/wagtail/wagtail/pull/10793 will be used as the base implementation for the Panel open/close behaviour also as this is mostly shared code.
